ledger/src/blockstore.rs

```diff
@@ -1416,32 +1416,23 @@ impl Blockstore {
 
         if !erasure_meta.check_coding_shred(&shred) {
             metrics.num_coding_shreds_invalid_erasure_config += 1;
-            let conflicting_shred = self.find_conflicting_coding_shred(
-                &shred,
-                slot,
-                erasure_meta,
-                just_received_shreds,
-            );
-            if let Some(conflicting_shred) = conflicting_shred {
-                if !self.has_duplicate_shreds_in_slot(slot) {
-                    if self
-                        .store_duplicate_slot(
-                            slot,
-                            conflicting_shred.clone(),
-                            shred.payload().clone(),
-                        )
-                        .is_err()
-                    {
-                        warn!("bad duplicate store..");
-                    }
+            if !self.has_duplicate_shreds_in_slot(slot) {
+                let conflicting_shred = self
+                    .find_conflicting_coding_shred(&shred, slot, erasure_meta, just_received_shreds)
+                    .expect("Shred indicated by erasure meta must exist")
```

Comment:
This `expect` here made me wonder what happens in the scenarios that a shred is purge/pruned/dumped from blockstore between the time `erasure_meta` is obtained until here the shred is fetched from blockstore?

Reply:
when replay decides to dump a block it uses `clear_unconfirmed_slot` which shares a lock with shred insertion precisely to avoid the situation you mentioned: 
https://github.com/anza-xyz/agave/blob/bf1b7651d4df9abde7963d59075f52011bee1dbe/ledger/src/blockstore.rs#L1290-L1291

i'm not an expert on the blockstore automatic cleanup (cc @steviez). while it doesn't share this lock, it looks at `blockstore.max_root()` in order to determine which slots to cleanup. We also compare against `blockstore.max_root()` earlier https://github.com/anza-xyz/agave/blob/bf1b7651d4df9abde7963d59075f52011bee1dbe/ledger/src/blockstore.rs#L1386
but this seems unsafe as `replay_stage` could have rooted and triggered an automatic cleanup in between.

Reply:
hmm, will recheck these blockstore expects under the assumption that cleanup could have happened https://discord.com/channels/428295358100013066/1235788902888116274/1235942333334290442

Reply:
You're correct that the `BlockstoreCleanupService` doesn't share the `insert_shreds_lock`. And right, the service fetches the latest root, and potentially clean data that is strictly older than the root. We never clean data newer than the latest root.

Practically speaking, the value for `--limit-ledger-size` is required to be `>= 50M`. Assuming 1k shreds per slot, the oldest data is ~50k slots older than the newest data. That being said, I think there is merit to an `unwrap()` and `expect()` audit like y'all are suggesting

Reply:
> hmm, will recheck these blockstore expects under the assumption that cleanup could have happened https://discord.com/channels/428295358100013066/1235788902888116274/1235942333334290442

Given that mainnet is upgrading to v1.18, and we have not have much of soak time, probably better to demote all (or most of) those `expect`s to an error log or metric. We can always add them back in if we are confident we are not missing an edge case.

Reply:
I combed through the expect/unwraps we introduced and demoted the affected ones here https://github.com/anza-xyz/agave/pull/1259 .

I believe these remaining checks to be safe and have not demoted them:
https://github.com/anza-xyz/agave/blob/fd8b0754ed22ccbe89709a38fd840a7249a794f8/ledger/src/blockstore.rs#L1134
https://github.com/anza-xyz/agave/blob/fd8b0754ed22ccbe89709a38fd840a7249a794f8/ledger/src/blockstore.rs#L1163